### PR TITLE
Single committee contract

### PR DESCRIPTION
**Pull request by vistra**
* Remove compliance committee contract, min-weight and min-committee-size related interfaces and tests
* Add compliance array to committee events
* Assign rewards on committee change

### CODE REVIEW

File: contracts/BootstrapRewards.sol [hunk 1]
```diff
@@ -65,35 +73,34 @@ contract BootstrapRewards is IBootstrapRewards, ContractRegistryAccessor {
...
+
             uint256 duration = now.sub(lastPayedAt);
-            uint256 amountPerValidator = Math.min(annualBootstrapRewards.mul(duration).div(365 days), pool.div(committee.length));
-            pool = pool.sub(amountPerValidator * committee.length);
+            uint256 amountPerGeneralValidator = Math.min(generalCommitteeAnnualBootstrap.mul(duration).div(365 days), pool.div(committee.length));
+            uint256 amountPerCompliantValidator = nCompliance == 0 ? 0 :
+                Math.min(complianceCommitteeAnnualBootstrap.mul(duration).div(365 days), pool.div(nCompliance));
 
-            uint256[] memory assignedRewards = new uint256[](committee.length);
+            pool = pool.sub(amountPerGeneralValidator * committee.length).sub(amountPerCompliantValidator * nCompliance);
```
Comment (vistra) [inline contracts/BootstrapRewards.sol:92]: Bug - should substract before second Math.min, or sum both elements and then do Math.min

File: contracts/Elections.sol [hunk 2]
```diff
@@ -126,51 +110,58 @@ contract Elections is IElections, ContractRegistryAccessor {
 		}
 	}
 
-	function isCommitteeVoteOutThresholdReached(address[] memory committee, uint256[] memory weights, address votee) private view returns (bool) {
+	function isCommitteeVoteOutThresholdReached(address[] memory committee, uint256[] memory weights, bool[] memory compliance, address votee) private view returns (bool) {
```
Comment (vistra) [inline contracts/Elections.sol:113]: Change to 2 invocations with a compliance flag

File: contracts/Elections.sol [hunk 3]
```diff
@@ -305,22 +303,19 @@ contract Elections is IElections, ContractRegistryAccessor {
...
 		}
-		getComplianceCommitteeContract().memberWeightChange(addr, getCommitteeEffectiveStake(addr));
-
 	}
 
 	function getCommitteeEffectiveStake(address v) private view returns (uint256) {
 		uint256 ownStake =  getStakingContract().getStakeBalanceOf(v);
-		bool isSelfDelegating = getDelegationsContract().getDelegation(v) == v; // TODO optimized three sequential calls to delegations in this function
-		if (!isSelfDelegating || ownStake == 0) {
+		if (ownStake == 0) {
```
Comment (vistra) [inline contracts/Elections.sol:314]: leave for s separate pr, add a todo instead

File: contracts/Fees.sol [hunk 4]
```diff
@@ -78,26 +84,40 @@ contract Fees is IFees, ContractRegistryAccessor {
...
+        }
+        if (n == 0) return;
+
         uint256 totalAssigned = 0;
 
         for (uint i = 0; i < committee.length; i++) {
-            uint256 curAmount = amount.div(committee.length);
-            assignedFees[i] = curAmount;
-            totalAssigned = totalAssigned.add(curAmount);
+            uint256 curAmount = amount.div(n);
```
Comment (vistra) [inline contracts/Fees.sol:106]: Do once outside the loop

File: contracts/Fees.sol [hunk 5]
```diff
@@ -78,26 +84,40 @@ contract Fees is IFees, ContractRegistryAccessor {
...
+                totalAssigned = totalAssigned.add(curAmount);
+            }
         }
 
         uint256 remainder = amount.sub(totalAssigned);
-        if (remainder > 0 && committee.length > 0) {
+        if (remainder > 0 && n > 0) {
             uint ind = now % committee.length;
+            if (isCompliant) {
+                while (!compliance[ind]) {
```
Comment (vistra) [inline contracts/Fees.sol:117]: This is not a fair draw - instead take `now % n` and find the n'th member.

File: contracts/StakingRewards.sol [hunk 6]
```diff
@@ -57,16 +64,15 @@ contract StakingRewards is IStakingRewards, ContractRegistryAccessor {
         return lastPayedAt;
     }
 
-    function assignRewards() external {
-        _assignRewards();
+    function assignRewards(address[] calldata committee, uint256[] calldata weights) external onlyElectionsContract {
+        _assignRewards(committee, weights);
     }
+    event GasReport(string label, uint gas);
```
Comment (vistra) [inline contracts/StakingRewards.sol:70]: clear

File: contracts/StakingRewards.sol [hunk 7]
```diff
@@ -57,16 +64,15 @@ contract StakingRewards is IStakingRewards, ContractRegistryAccessor {
...
 
-    function _assignRewards() private {
+    function _assignRewards(address[] memory committee, uint256[] memory weights) private {
         // TODO we often do integer division for rate related calculation, which floors the result. Do we need to address this?
         // TODO for an empty committee or a committee with 0 total stake the divided amounts will be locked in the contract FOREVER
-
-        (address[] memory committee, uint256[] memory weights) = getGeneralCommitteeContract().getCommittee();
-
+        uint g = gasleft();
```
Comment (vistra) [inline contracts/StakingRewards.sol:75]: clear

File: contracts/StakingRewards.sol [hunk 8]
```diff
@@ -77,36 +83,33 @@ contract StakingRewards is IStakingRewards, ContractRegistryAccessor {
...
-                assignedRewards[i] = curAmount;
-                totalAssigned = totalAssigned.add(curAmount);
+                n = amount.mul(weights[i]).div(totalWeight);
+                assignedRewards[i] = n;
+                totalAssigned = totalAssigned.add(n);
             }
 
-            uint256 remainder = amount.sub(totalAssigned);
-            if (remainder > 0 && committee.length > 0) {
+            n = amount.sub(totalAssigned);
```
Comment (vistra) [inline contracts/StakingRewards.sol:97]: check if can use proper names here

File: contracts/spec_interfaces/IBootstrapRewards.sol [hunk 9]
```diff
@@ -5,7 +5,7 @@ import "./IContractRegistry.sol";
 
 /// @title Fees contract interface
 interface IBootstrapRewards {
-    event BootstrapRewardsAssigned(address[] assignees, uint256[] amounts);
+    event BootstrapRewardsAssigned(address[] assignees, uint256[] amounts, uint256 sinceTime);
```
Comment (vistra) [inline contracts/spec_interfaces/IBootstrapRewards.sol:8]: remove sinceTime

File: test/event-parsing.ts [hunk 10]
```diff
@@ -59,3 +59,7 @@ export const banningVoteEvents = (txResult, contractAddress?: string) => parseLo
 export const electionsBanned = (txResult, contractAddress?: string) => parseLogs(txResult, elections, "Banned(address)", contractAddress);
 export const electionsUnbanned = (txResult, contractAddress?: string) => parseLogs(txResult, elections, "Unbanned(address)", contractAddress);
 export const validatorComplianceUpdateEvents = (txResult, contractAddress?: string) => parseLogs(txResult, compliance, "ValidatorComplianceUpdate(address,bool)", contractAddress);
+
+export const gasReportEvents = (txResult, contractAddress?: string) => parseLogs(txResult, committee, "GasReport(string,uint256)", contractAddress);
```
Comment (vistra) [inline test/event-parsing.ts:63]: clean

File: test/gas-scenarios.spec.ts [hunk 11]
```diff
@@ -23,48 +24,55 @@ const BASE_STAKE = 1000000;
...
+    await Promise.all(_.shuffle(committee.concat(standbys)).map(v => v.notifyReadyForCommittee()));
 
     const monthlyRate = 1000;
     const subs = await d.newSubscriber('defaultTier', monthlyRate);
     const appOwner = d.newParticipant();
 
     for (let i = 0; i < numVCs; i++) {
         await createVC(d, false, subs, monthlyRate, appOwner);
-        await createVC(d, true, subs, monthlyRate, appOwner);
+        // await createVC(d, true, subs, monthlyRate, appOwner);
```
Comment (vistra) [inline test/gas-scenarios.spec.ts:73]: remove comment

File: test/gas-scenarios.spec.ts [hunk 12]
```diff
@@ -75,18 +83,26 @@ async function fullCommitteeAndStandbys(committeeEvenStakes:boolean = false, sta
 
 
 describe('gas usage scenarios', async () => {
-    it("New delegator stake increase, lowest committee member gets to top", async () => {
+    it.only("New delegator stake increase, lowest committee member gets to top", async () => {
```
Comment (vistra) [inline test/gas-scenarios.spec.ts:86]: remove onlys

File: test/gas-scenarios.spec.ts [hunk 13]
```diff
@@ -75,18 +83,26 @@ async function fullCommitteeAndStandbys(committeeEvenStakes:boolean = false, sta
...
+        await evmIncreaseTime(d.web3, 30*24*60*60);
+
         d.resetGasRecording();
         let r = await delegator.stake(BASE_STAKE * 1000);
         expect(r).to.have.a.committeeChangedEvent({
             addrs: committee.map(v => v.address)
         });
 
+        // const ge = gasReportEvents(r).concat(gasReportEvents2(r));
+        const ge = gasReportEvents2(r);
```
Comment (vistra) [inline test/gas-scenarios.spec.ts:103]: clean?

File: test/gas-scenarios.spec.ts [hunk 14]
```diff
@@ -377,31 +393,56 @@ describe('gas usage scenarios', async () => {
...
+        // await d.bootstrapRewards.assignRewards({from: p.address});
+        // d.logGasUsageSummary("assigns rewards (bootstrap) (1 month, initial balance == 0)", [p]);
+        // d.resetGasRecording();
+        // await d.fees.assignFees({from: p.address});
+        // d.logGasUsageSummary("assigns rewards (fees) (1 month, initial balance == 0)", [p]);
 
         d.logGasUsageSummary("assigns rewards (1 month, initial balance > 0)", [p]);
     });
 
+    it("test", async () => {
```
Comment (vistra) [inline test/gas-scenarios.spec.ts:438]: clean

File: test/gas-scenarios.spec.ts [hunk 15]
```diff
@@ -377,31 +393,56 @@ describe('gas usage scenarios', async () => {
...
-    it("assigns rewards (1 month, initial balance > 0)", async () => {
-        const {d, committee, standbys} = await fullCommitteeAndStandbys(false, false, 100);
+    it.only("assigns rewards (1 month, initial balance > 0)", async () => {
+        const {d, committee, standbys} = await fullCommitteeAndStandbys(false, false, 5);
         await evmIncreaseTime(d.web3, 30*24*60*60);
 
-        await d.stakingRewards.assignRewards();
-        await d.bootstrapRewards.assignRewards();
-        await d.fees.assignFees();
+        // await d.stakingRewards.assignRewards();
```
Comment (vistra) [inline test/gas-scenarios.spec.ts:414]: clean commented-out code

File: test/staking-rewards.spec.ts [hunk 16]
```diff
@@ -162,14 +161,15 @@ describe('staking-rewards-level-flows', async () => {
...
+    r = await v2.notifyReadyForCommittee();
+    const startTime = await txTimestamp(d.web3, r);
 
     const validators = [{
-      v: v2,
-      stake: initStakeLarger
-    }, {
       v: v1,
       stake: initStakeLesser
+    }, {
```
Comment (vistra) [inline test/staking-rewards.spec.ts:170]: did the order matter?

File: contracts/Committee.sol [hunk 17]
```diff
@@ -8,13 +8,16 @@ import "./ContractRegistryAccessor.sol";
 
 /// @title Elections contract interface
 contract Committee is ICommittee, ContractRegistryAccessor {
+	event GasReport(string label, uint gas);
```
Comment (vistra) [inline contracts/Committee.sol:11]: clean

File: contracts/Committee.sol [hunk 18]
```diff
@@ -27,35 +30,30 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-		uint minStandbyWeight;
-		address minStandbyAddress;
-		uint standbysCount;
-
-		// Committee entry barrier
-		uint committeeSize;
-		uint minCommitteeMemberWeight;
+	// Derived properties
+	struct CommitteeInfo {
 		address minCommitteeMemberAddress;
```
Comment (vistra) [inline contracts/Committee.sol:52]: Is it still needed?

File: contracts/Committee.sol [hunk 19]
```diff
@@ -27,35 +30,30 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+		uint64 committeeBitmap;
```
Comment (vistra) [inline contracts/Committee.sol:53]: The 64 length limit of the participants array needs to be enforced

File: contracts/Committee.sol [hunk 20]
```diff
@@ -27,35 +30,30 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+		uint8 standbysCount;
+		uint8 committeeSize;
+		bool pendingChanges;
```
Comment (vistra) [inline contracts/Committee.sol:56]: Should be unused, remove

File: contracts/Committee.sol [hunk 21]
```diff
@@ -27,35 +30,30 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+		uint32 readyToSyncTimeout;
+		uint8 maxCommitteeSize;
+		uint8 maxStandbys;
 	}
 	Settings settings;
 
-	// Derived properties (TODO can be reduced to 2-3 state entries)
-	struct CommitteeInfo {
-		uint freeParticipantSlotPos;
+	uint256 weightSortIndicesBytes;
```
Comment (vistra) [inline contracts/Committee.sol:48]: This would only allow at most 32 participants, probably needs to be more

File: contracts/Committee.sol [hunk 22]
```diff
@@ -86,28 +81,28 @@ contract Committee is ICommittee, ContractRegistryAccessor {
 
 	/// @dev Called by: Elections contract
 	/// Notifies a weight change for sorting to a relevant committee member.
-    /// weight = 0 indicates removal of the member from the committee (for example on unregister, voteUnready, voteOut)
+	/// weight = 0 indicates removal of the member from the committee (for example on unregister, voteUnready, voteOut)
 	function memberWeightChange(address addr, uint256 weight) external onlyElectionsContract returns (bool committeeChanged, bool standbysChanged) {
 		MemberData memory memberData = membersData[addr];
 		if (memberData.isMember) {
-			memberData.weight = weight;
+			memberData.weight = uint128(weight);
```
Comment (vistra) [inline contracts/Committee.sol:88]: require weight is in proper range

File: contracts/Committee.sol [hunk 23]
```diff
@@ -8,13 +8,16 @@ import "./ContractRegistryAccessor.sol";
 
 /// @title Elections contract interface
 contract Committee is ICommittee, ContractRegistryAccessor {
+	event GasReport(string label, uint gas);
+
 	address[] participants;
 
 	struct MemberData { // TODO can be reduced to 1 state entry
+		uint128 weight;
+		uint32 readyToSyncTimestamp;
```
Comment (vistra) [inline contracts/Committee.sol:17]: change to uint64

File: contracts/Committee.sol [hunk 24]
```diff
@@ -117,6 +112,18 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+				addr: addr,
+				data: memberData
+				}));
+		}
+		return (false, false);
+	}
+
+	function memberComplianceChange(address addr, bool isCompliant) external onlyElectionsContract returns (bool commiteeChanged, bool standbysChanged) {
+		MemberData memory memberData = membersData[addr];
+		if (memberData.isMember) {
```
Comment (vistra) [inline contracts/Committee.sol:125]: fix indentations

File: contracts/Committee.sol [hunk 25]
```diff
@@ -175,44 +175,68 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+
+		addrs = new address[](committeeSize);
+		weights = new uint[](committeeSize);
+		compliance = new bool[](committeeSize);
+		bitmap = uint(_committeeInfo.committeeBitmap);
+		uint i = 0;
+		MemberData memory md;
+		while (bitmap != 0) {
+			if (bitmap & 1 == 1) {
+				addrs[i] = participants[i];
```
Comment (vistra) [inline contracts/Committee.sol:192]: BUG - participants[i] should use a different index that advances with each loop iteration

File: contracts/Committee.sol [hunk 26]
```diff
@@ -175,44 +175,68 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+		uint bitmap = uint(_committeeInfo.committeeBitmap);
+		uint committeeSize = _committeeInfo.committeeSize;
+
+		addrs = new address[](committeeSize);
+		weights = new uint[](committeeSize);
+		compliance = new bool[](committeeSize);
+		bitmap = uint(_committeeInfo.committeeBitmap);
+		uint i = 0;
+		MemberData memory md;
+		while (bitmap != 0) {
```
Comment (vistra) [inline contracts/Committee.sol:190]: maybe use a for loop instead

File: contracts/Committee.sol [hunk 27]
```diff
@@ -175,44 +175,68 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-		_settings.minimumWeight = _minimumWeight;
-		_settings.minimumAddress = _minimumAddress;
-		_settings.minCommitteeSize = _minCommitteeSize;
-		settings = _settings;
-
-		updateOnMemberChange(NullMember(), _settings);
+		addrs = new address[](standbysCount);
+		weights = new uint[](standbysCount);
+		bitmap = uint(_committeeInfo.committeeBitmap);
+		uint i = 0;
```
Comment (vistra) [inline contracts/Committee.sol:211]: better names

File: contracts/Committee.sol [hunk 28]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
 
 		bool isParticipant = member.data.inCommittee || member.data.isStandby;
 		if (!isParticipant) {
-			CommitteeInfo memory _committeeInfo = committeeInfo;
-			if (!qualifiedToJoin(member, _settings, _committeeInfo)) {
+			if (!member.data.isMember || member.data.readyToSyncTimestamp == 0 || member.data.weight == 0) {
```
Comment (vistra) [inline contracts/Committee.sol:272]: Use qualifiesAsStandby instead

File: contracts/Committee.sol [hunk 29]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-
-					isStandby: true,
-					inCommittee: false
-				})
-			}),
-		_settings) > 0;
-	}
-
-	function qualifiesForCommittee(Member memory member, Settings memory _settings, uint committeeSize, bool _outranksLowestCommitteeMember) private view returns (bool) {
+	function qualifiesForCommittee(Participant memory member, Settings memory _settings, uint committeeSize) private view returns (bool) {
```
Comment (vistra) [inline contracts/Committee.sol:288]: better name? this is used for construction

File: contracts/Committee.sol [hunk 30]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-		participants.length = Math.max(participants.length, _committeeInfo.freeParticipantSlotPos + 1);
-		participants[_committeeInfo.freeParticipantSlotPos] = member.addr;
+	struct UpdateVars {
+		uint maxPos;
+		Participant p;
+		uint seenStandbys;
+		uint seenCommittee;
 	}
-
 	function updateOnMemberChange(Member memory member, Settings memory _settings) private returns (bool committeeChanged, bool standbysChanged) { // TODO this is sometimes called with a member with address 0 indicating no member changed
```
Comment (vistra) [inline contracts/Committee.sol:304]: does the todo still hold?

File: contracts/Committee.sol [hunk 31]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
 	}
 
-	function qualifiedToJoinCommittee(Member memory member, Settings memory _settings, CommitteeInfo memory _committeeInfo) private view returns (bool) {
-		return qualifiesForCommittee(member, _settings, _committeeInfo.committeeSize, outranksLowestCommitteeMember(member, _committeeInfo, _settings));
-	}
-
-	function joinMemberAsParticipant(Member memory member, CommitteeInfo memory _committeeInfo) private {
-		participants.length = Math.max(participants.length, _committeeInfo.freeParticipantSlotPos + 1);
-		participants[_committeeInfo.freeParticipantSlotPos] = member.addr;
+	struct UpdateVars {
```
Comment (vistra) [inline contracts/Committee.sol:298]: still needed?

File: contracts/Committee.sol [hunk 32]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-		if (committeeChanged) _notifyCommitteeChanged(o.committee);
-		if (standbysChanged) _notifyStandbysChanged(o.standbys);
-	}
+		(Participant[] memory _members, Participant memory memberAsParticipant) = loadParticipantsSortedByWeights(_participants, member); // override stored member with preloaded one //70k
+
+		CommitteeInfo memory newCommitteeInfo;
+
+		UpdateVars memory s;
+		Participant memory p;
+		uint newWeightSortIndicesBytes;
```
Comment (vistra) [inline contracts/Committee.sol:315]: make explicit uint256

File: contracts/Committee.sol [hunk 33]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+		bool changed;
+		for (uint i = 0; i < _members.length; i++) {
+			p = _members[i];
+			if (qualifiesForCommittee(p, _settings, newCommitteeInfo.committeeSize)) {
+				p.shouldBeInCommittee = true;
+				newCommitteeInfo.committeeSize++;
+				newCommitteeInfo.committeeBitmap |= uint64(uint(1) << p.pos);
+				newCommitteeInfo.minCommitteeMemberAddress = p.addr; // list is sorted by weight in descending order
```
Comment (vistra) [inline contracts/Committee.sol:323]: do we need this? maybe remove

File: contracts/Committee.sol [hunk 34]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+			} else if (
+				newCommitteeInfo.standbysCount < _settings.maxStandbys &&
```
Comment (vistra) [inline contracts/Committee.sol:325]: maybe wrap in a function?

File: contracts/Committee.sol [hunk 35]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-		ci.freeParticipantSlotPos = Math.min(ci.freeParticipantSlotPos, r.minPos);
-
-		participants.length = Math.min(_participants.length, maxPos + 1);
-
-		committeeInfo = ci;
-
-		if (committeeChanged) _notifyCommitteeChanged(o.committee);
-		if (standbysChanged) _notifyStandbysChanged(o.standbys);
-	}
+		(Participant[] memory _members, Participant memory memberAsParticipant) = loadParticipantsSortedByWeights(_participants, member); // override stored member with preloaded one //70k
```
Comment (vistra) [inline contracts/Committee.sol:309]: rename to _participants

File: contracts/Committee.sol [hunk 36]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-			if (p.data.isStandby != isStandby) {
-				p.data.isStandby = isStandby;
-				if (isStandby && p.data.inCommittee) {
-					p.data.readyToSyncTimestamp = now;
+		for (uint i = 0; i < _members.length; i++) {
+			p = _members[i];
+			changed = false;
+			if (p.shouldBeStandby != p.data.isStandby) {
+				if (
+					!p.shouldBeInCommittee && !p.shouldBeStandby &&
```
Comment (vistra) [inline contracts/Committee.sol:339]: use p.data.isStandby instead of !p.shouldBeStandby

File: contracts/Committee.sol [hunk 37]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+					newCommitteeInfo.standbysCount < _settings.maxStandbys &&
+					qualifiesAsStandby(p)
+				) {
+					p.shouldBeStandby = true;
+					newCommitteeInfo.standbysCount++;
+				}
+				if (p.shouldBeStandby != p.data.isStandby) {
+					p.data.isStandby = p.shouldBeStandby;
+					if (p.shouldBeStandby && p.data.inCommittee) {
```
Comment (vistra) [inline contracts/Committee.sol:348]: add comment to explain this

File: contracts/Committee.sol [hunk 38]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+			if (changed) {
 				membersData[p.addr] = p.data;
 			}
+			if (!p.data.inCommittee && !p.data.isStandby) {
+				// no longer a participant
+				if (p.pos < _participants.length) {
+					participants[p.pos] = address(0);
+				}
+			} else {
+				newWeightSortIndicesBytes = (newWeightSortIndicesBytes << 8) | uint8(p.pos + 1);
```
Comment (vistra) [inline contracts/Committee.sol:369]: Maybe rename to emphasise that indices are one-based

File: contracts/Committee.sol [hunk 39]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-			r.minTimestamp = Math.min(r.minTimestamp, p.data.readyToSyncTimestamp);
-			if (p.data.weight < r.minWeight) {
-				r.minWeight = p.data.weight;
-				r.minAddr = p.addr;
-			} else if (p.data.weight == r.minWeight && uint(p.addr) < uint(r.minAddr)) {
-				r.minAddr = p.addr;
+		// check if member is a new participant
+		if (
+			(memberAsParticipant.data.inCommittee || memberAsParticipant.data.isStandby) &&
+			(_participants.length == memberAsParticipant.pos || _participants[memberAsParticipant.pos] != memberAsParticipant.addr)
```
Comment (vistra) [inline contracts/Committee.sol:377]: can be replaced with comparison to zero address

File: contracts/Committee.sol [hunk 40]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-				nParticipants++;
-			} else if (firstFreeSlot == participantsAddrs.length) {
-				firstFreeSlot = i;
+		weightSortIndicesBytes = newWeightSortIndicesBytes;
+		committeeInfo = newCommitteeInfo; // todo check if changed before writing
+
+		notifyChanges(_members, newCommitteeInfo.committeeSize, newCommitteeInfo.standbysCount, committeeChanged, standbysChanged);
+	}
+
+	function notifyChanges(Participant[] memory members, uint committeeSize, uint standbysCount, bool committeeChanged, bool standbysChanged) private {
```
Comment (vistra) [inline contracts/Committee.sol:396]: don't allocate and build unnecessary arrays if no need to emit events

File: contracts/Committee.sol [hunk 41]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
 
-		_participants = new Participant[](nParticipants);
-		uint mInd = 0;
+		if (committeeChanged) _notifyCommitteeChanged(committeeAddrs, committeeWeights, committeeCompliance);
+		if (standbysChanged) _notifyStandbysChanged(standbyAddrs, standbyWeights, standbysCompliance);
+	}
+
+	function loadParticipantsSortedByWeights(address[] memory participantsAddrs, Member memory preloadedMember) private returns (Participant[] memory _participants, Participant memory memberAsParticipant) {
+		uint nParticipants;
+		bool foundMember = preloadedMember.data.inCommittee || preloadedMember.data.isStandby;
```
Comment (vistra) [inline contracts/Committee.sol:426]: newMember instead of foundMember

File: contracts/Committee.sol [hunk 42]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
+		uint preloadedInd = uint(-1);
+		uint preloadedPos = uint(-1);
+		while (sortBytes != 0) {
+			pos = uint(sortBytes & 0xFF) - 1;
+			addr = participantsAddrs[pos];
+			if (addr != preloadedMember.addr) {
+				md = membersData[addr];
+				if (
+					preloadedInd == uint(-1) && (
+					md.weight > preloadedMember.data.weight || (md.weight == preloadedMember.data.weight && uint(addr) > uint(preloadedMember.addr)))
```
Comment (vistra) [inline contracts/Committee.sol:455]: extract to a separate function?

File: contracts/Committee.sol [hunk 43]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-		uint nStandbys = 0;
-		while (
-			i < _participants.length &&
-			nStandbys < _settings.maxStandbys &&
-			qualifiesAsStandby(Member({addr: _participants[i].addr, data: _participants[i].data}))
-		) {
-			nStandbys++;
-			i++;
+		if (preloadedInd == uint(-1)) {
+			preloadedInd = pind;
```
Comment (vistra) [inline contracts/Committee.sol:476]: pind == 0 by now, so use 0 instead

File: contracts/Committee.sol [hunk 44]
```diff
@@ -264,446 +269,228 @@ contract Committee is ICommittee, ContractRegistryAccessor {
...
-            }
-        }
-	}
-
-	function compareMembers(Participant memory a, Participant memory b, Settings memory _settings, Comparator comparator) private view returns (int) {
-		Member memory ma = Member({addr: a.addr, data: a.data});
-		Member memory mb = Member({addr: b.addr, data: b.data});
-		if (comparator == Comparator.Committee) {
-			return compareMembersPerCommitteeCriteria(ma, mb, _settings);
+		if (preloadedPos != uint(-1)) {
```
Comment (vistra) [inline contracts/Committee.sol:482]: maybe newMember instead?

File: contracts/Committee.sol (hunk 19 above)
Comment (vistra) [reply contracts/Committee.sol:50]: actually it's 32

File: contracts/Committee.sol [hunk 45]
```diff
@@ -83,71 +80,79 @@ contract Committee is ICommittee, ContractRegistryAccessor {
 	/// Notifies a weight change for sorting to a relevant committee member.
 	/// weight = 0 indicates removal of the member from the committee (for example on unregister, voteUnready, voteOut)
 	function memberWeightChange(address addr, uint256 weight) external onlyElectionsContract returns (bool committeeChanged, bool standbysChanged) {
+		require(uint256(uint128(weight)) == weight, "weight is out of range");
```
Comment (ronnno) [inline contracts/Committee.sol:83]: consider checking the value is below a constant instead of casting.

File: contracts/Committee.sol [hunk 46]
```diff
@@ -8,13 +8,11 @@ import "./ContractRegistryAccessor.sol";
...
 contract Committee is ICommittee, ContractRegistryAccessor {
-	event GasReport(string label, uint gas);
-
-	address[] participants;
+	address[] participantAddresses;
 
 	struct MemberData { // TODO can be reduced to 1 state entry
 		uint128 weight;
-		uint32 readyToSyncTimestamp;
+		uint48 readyToSyncTimestamp;
```
[the remark this answers is not in the document]
Comment (vistra) [reply contracts/Committee.sol:15]: 9 million years :)

File: contracts/BootstrapRewards.sol [hunk 47]
```diff
@@ -86,18 +86,19 @@ contract BootstrapRewards is IBootstrapRewards, ContractRegistryAccessor {
 
             uint256 duration = now.sub(lastPayedAt);
             uint256 amountPerGeneralValidator = Math.min(generalCommitteeAnnualBootstrap.mul(duration).div(365 days), pool.div(committee.length));
+            pool = pool.sub(amountPerGeneralValidator * committee.length);
```
[the remark this answers is not in the document]
Comment (vistra) [reply contracts/BootstrapRewards.sol:89]: It's just 2, not times committee length. Also, it will be removed entirely in a future PR because we will not check the pool at this point but only on withdrawal.
